File: core/core-application/src/test/java/com/nmh/commerce/coupon/CouponStockConcurrencyTest.java
package com.nmh.commerce.coupon;

import com.nmh.commerce.CoreApplicationIntegrationTest;
import com.nmh.commerce.domain.Quantity;
import org.junit.jupiter.api.Test;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static org.assertj.core.api.BDDAssertions.then;

public class CouponStockConcurrencyTest extends CoreApplicationIntegrationTest {
    private final CouponStockRepository couponStockRepository;
    private final CouponStockManager couponStockManager;

    public CouponStockConcurrencyTest(CouponStockRepository couponStockRepository, CouponStockManager couponStockManager) {
        this.couponStockRepository = couponStockRepository;
        this.couponStockManager = couponStockManager;
    }

    @Test
    void 동시에_100개의_쿠폰의_재고를_감소시킨다() throws InterruptedException {
        // given
        CouponStock savedStock = couponStockRepository.save(CouponStock.of(1L, Quantity.of(10)));

        int threadCount = 10;
        ExecutorService executorService = Executors.newFixedThreadPool(2);

        CountDownLatch latch = new CountDownLatch(threadCount);
        // when
        for (int i = 0; i < threadCount; i++) {
            executorService.execute(() -> {
                try {
                    couponStockManager.deductStock(savedStock.getCouponId());
                } finally {
                    latch.countDown();
                }
            });
        }

        latch.await();

        CouponStock stock = couponStockRepository.findByCouponId(savedStock.getCouponId()).orElseThrow();
        // then
        then(stock.getRemainingQuantity().value).isEqualTo(0);
    }
}

File: core/core-application/src/test/kotlin/com/nmh/commerce/coupon/CouponStockConcurrencyTest.kt
package com.nmh.commerce.coupon

import com.nmh.commerce.CoreApplicationIntegrationTest
import com.nmh.commerce.domain.Quantity.Companion.of
import org.assertj.core.api.BDDAssertions
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors

class CouponStockConcurrencyTest(
    private val couponStockRepository: CouponStockRepository,
    private val couponStockManager: CouponStockManager,
) : CoreApplicationIntegrationTest() {
    @Disabled("잠시 비활성화")
    @Test
    @Throws(InterruptedException::class)
    fun 동시에_100개의_쿠폰의_재고를_감소시킨다() {
        // given
        val couponStock = CouponStock(1L, 1L, of(10), null)
        val savedStock = couponStockRepository.save(couponStock)

        val threadCount = 10
        val executorService = Executors.newFixedThreadPool(2)

        val latch = CountDownLatch(threadCount)
        // when
        for (i in 0..<threadCount) {
            executorService.execute(
                Runnable {
                    try {
                        couponStockManager.deductStock(savedStock.couponId)
                    } finally {
                        latch.countDown()
                    }
                },
            )
        }

        latch.await()

        val stock: CouponStock = couponStockRepository.findByCouponId(savedStock.couponId)
        // then
        BDDAssertions.then(stock.remainingQuantity.value).isEqualTo(0)
    }
}
